### PR TITLE
Make "having microkernels" a target attribute.

### PR DESCRIPTION
**Pull request by bjacob**
We need that information in `MaterializeEncodingPass`, as we need to switch to dynamic tile sizes in that case.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 1]
```diff
@@ -123,7 +133,12 @@ bool isRISCV(IREE::HAL::ExecutableTargetAttr targetAttr) {
 }
 
 bool isVMVXBackend(IREE::HAL::ExecutableTargetAttr targetAttr) {
-  return targetAttr.getBackend().getValue().startswith("vmvx");
+  return targetAttr && targetAttr.getBackend().getValue().startswith("vmvx");
+}
+
+bool hasMicrokernels(IREE::HAL::ExecutableTargetAttr targetAttr) {
+  auto enableMicrokernels = getConfigBoolAttr(targetAttr, "ukernels");
+  return enableMicrokernels && enableMicrokernels->getValue();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:141]: nit: use `.value()`

I recently notice that there is a plan about deprecating `Optional<X>::hasValue/getValue/getValueOr`. I think we should start using the new style.

https://discourse.llvm.org/t/deprecating-llvm-optional-x-hasvalue-getvalue-getvalueor/63716/15

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 2]
```diff
@@ -92,6 +92,16 @@ Optional<IntegerAttr> getConfigIntegerAttr(
   return attr;
 }
 
+Optional<BoolAttr> getConfigBoolAttr(IREE::HAL::ExecutableTargetAttr targetAttr,
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:95]: We should put it to an unnamed namespace, or add `static` to it.

**Review by hanhanW — APPROVED** [review]
overall looks good to me, just few nits!

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 1 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:141]: Ah, but here in the expression `enableMicrokernels->getValue()`, it is the `operator->` that queries the `Optional::value()` already, returning a `BoolAttr`, so the `getValue()` call here is `BoolAttr::getValue`. I checked if `BoolAttr` was undergoing a similar renaming but it seems not to:

```
/usr/local/google/home/benoitjacob/iree/compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:141:52: error: no member named 'value' in 'mlir::BoolAttr'
  return enableMicrokernels && enableMicrokernels->value();
```

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 2 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:95]: Good point! But the above `getConfigStringAttr` and `getConfigIntegerAttr` are already public, and already have one non-local caller each. So I wanted this function to be consistent with that, even though it did not not already have a non-local caller. I did forget to add its declaration to `Utils.h` - fixed, thanks for catching that.

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 1 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:141]: @MaheshRavishankar this is an example of a conversation that has been had due to the nesting of attributes in `Optional` :wink:   (insider joke / see #11301)

**Pull request merged**